Phlex HerbariumForm component

nimmolo opened the pull request:
## Manual tests requested

- **Herbarium form**: new and edit. 
- **Observation search: names field**. 
Because this PR refactors the "append" prop of the `AutocompleterField` component into a "slot" for easier HTML handling, please test the Name fields in the obs (or name) search forms too. 

Moves the "herbarium code" help text fully into the help text, to tidy up the form. 
Currently on main:
![Screen Shot 2025-12-28 at 10 58 54 PM](https://github.com/user-attachments/assets/9e54eaf7-fc09-455f-a55c-dd31acd8dcca)
On this branch:
![Screen Shot 2025-12-28 at 10 58 16 PM](https://github.com/user-attachments/assets/6afb648a-2108-44f7-9c6e-3633f0d73c85)
Expanded help. If anyone's wondering what this is, i have to think they're going to click the (?).
![Screen Shot 2025-12-28 at 10 58 39 PM](https://github.com/user-attachments/assets/86e5f121-c97e-436a-8312-40de5ba2fbf7)


New Components:
  - app/components/herbarium_form.rb - Main Phlex/Superform form component
  - app/components/form_location_map.rb - Map component with toggle/clear buttons
  - app/components/bounds_hidden_fields.rb - Hidden fields for location bounds

  New Tests:
  - test/components/herbarium_form_test.rb (20 tests)
  - test/components/form_location_map_test.rb (10 tests)
  - test/components/bounds_hidden_fields_test.rb (5 tests)

  Modified Files:
  - app/components/application_form.rb - Added :between to WRAPPER_OPTIONS
  - app/components/application_form/autocompleter_field.rb - Added support for between: option
  - app/controllers/herbaria_controller.rb - Updated modal form rendering
  - app/helpers/modals_helper.rb - Added Herbarium to component_map
  - app/views/controllers/herbaria/new.html.erb - Uses new component
  - app/views/controllers/herbaria/edit.html.erb - Uses new component
  - test/controllers/herbaria_controller_test.rb - Added turbo_stream tests

nimmolo commented on app/components/application_form/autocompleter_field.rb:
```diff
@@ -17,10 +19,16 @@ class AutocompleterField < Superform::Rails::Components::Input
+
+    # Make slot accessor public (Phlex::Slotable makes them private by default)
+    public :append_slot
+
```
@mo-nathan The reason for this change is that in Phlex, a `slot` rather than a `prop` is the best way to handle a "component arg" that contains HTML. If it's a `slot` it's got to be valid HTML; if it's a `prop` it's typed as a `String` but unvalidated as HTML.

The `:append` slot is for HTML that may be printed below the autocompleter field.

mo-nathan reviewed (changes requested):
The Copilot comments seem reasonable to address to me.  Otherwise this looks good.  Manual testing went fine.

nimmolo commented on app/components/herbarium_form.rb:
```diff
@@ -0,0 +1,208 @@
+      raw(:edit_herbarium_user_records.t(
+        name: "#{name} (#{login})", num: count
+      ).html_safe)
+      # rubocop:enable Rails/OutputSafety
```
Copilot is incorrect here. `name` and `login` are not user-supplied content, they are user attributes straight out of the db.

JoeCohen commented:
Pre-existing Errors (probably outside the scope of this PR):
- http://localhost:3000/herbaria/new
- Fill in Name with a new name
- Fill in Standard Abbreviation with `F`
- Create
- Result:
```ruby
ActionController::UrlGenerationError in HerbariaController#create
No route matches {:action=>"show", :controller=>"herbaria", :id=>#<Herbarium id: nil, mailing_address: "", location_id: nil, email: "", name: "jdc test", description: "", created_at: nil, updated_at: nil, code: "F", personal_user_id: nil>}, possible unmatched constraints: [:id]

Did you mean?
herbarium_url

Extracted source (around line #413):
    respond_to do |format|
      format.html do
        redirect_to(herbarium_path(@herbarium)) and return
      end
      format.turbo_stream do
        # Context here is the obs form.

Rails.root: /Users/joe/mushroom-observer

Application Trace | Framework Trace | Full Trace
app/controllers/herbaria_controller.rb:413:in `block (2 levels) in show_modal_flash_or_show_herbarium'
app/controllers/herbaria_controller.rb:411:in `show_modal_flash_or_show_herbarium'
app/controllers/herbaria_controller.rb:388:in `redirect_to_create_location_or_referrer_or_show_location'
app/controllers/herbaria_controller.rb:205:in `create'
app/controllers/application_controller.rb:148:in `catch_errors_and_log_request_stats'
Request
Parameters:

{"authenticity_token"=>"[FILTERED]",
 "_method"=>"post",
 "herbarium"=>{"name"=>"jdc test", "personal"=>"0", "code"=>"F", "place_name"=>"", "location_id"=>"", "email"=>"", "mailing_address"=>"", "description"=>""},
 "location"=>{"north"=>"", "south"=>"", "east"=>"", "west"=>"", "low"=>"", "high"=>""},
 "commit"=>"Create"}
```

JoeCohen commented:
Thanks for the fixes.
If it were a priority and we had the time, we could probably find more
issues (outside the scope of this PR) with herbaria.
Herbaria really needs an overall, but I think that's for after Bootstrap.

On Wed, Dec 31, 2025 at 1:15 PM andrew nimmo ***@***.***>
wrote:

> *nimmolo* left a comment (MushroomObserver/mushroom-observer#3640)
> <https://github.com/MushroomObserver/mushroom-observer/pull/3640#issuecomment-3702912828>
>
> Fixes #3670
> <https://github.com/MushroomObserver/mushroom-observer/issues/3670>
> Fixes #3671
> <https://github.com/MushroomObserver/mushroom-observer/issues/3671>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/3640#issuecomment-3702912828>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFFF7LO6GBHXULIBTVD4EQ4IXAVCNFSM6AAAAACQENS6C6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTOMBSHEYTEOBSHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

nimmolo commented:
Summary of fixes:

1. Save Failure Handling (`ActionController::UrlGenerationError` fix)
File: `app/controllers/herbaria_controller.rb`
  - Added `unless @herbarium.save` check in both create and update actions
  - When save fails, now properly calls `flash_object_errors(@herbarium)` and renders the form instead of trying to redirect with a nil ID

2. Blank Code Unique Constraint (`ActiveRecord::RecordNotUnique` fix)
File: `app/controllers/herbaria_controller.rb`
  - Added `@herbarium.code = @herbarium.code.presence` in `normalize_parameters`
  - This converts empty strings to `nil`, which the database's partial unique index correctly allows multiple of

3. Updated Tests for Nil Code
File: `test/controllers/herbaria_controller_test.rb`
  - Changed 2 assertions from `assert_equal("", herbarium.code)` to `assert_nil(herbarium.code)`

4. Added Regression Tests
File: `test/controllers/herbaria_controller_test.rb`
  - `test_create_with_blank_code_stores_nil` - verifies multiple herbaria can have blank codes (stored as nil)
  - `test_create_save_failure_returns_to_form` - verifies save failure shows form, not a redirect error
